x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/rules_feature_tour_context.tsx

```diff
@@ -0,0 +1,156 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { createContext, useCallback, useContext, useEffect, useMemo, FC } from 'react';
+
+import { noop } from 'lodash';
+import {
+  useEuiTour,
+  EuiStatelessTourStep,
+  EuiSpacer,
+  EuiButton,
+  EuiTourStepProps,
+  EuiTourStep,
+} from '@elastic/eui';
+import { invariant } from '../../../../../../common/utils/invariant';
+import { useKibana } from '../../../../../common/lib/kibana';
+
+import * as i18n from '../translations';
+
+export interface RulesFeatureTourContextType {
+  steps: {
+    inMemoryTableStepProps: EuiTourStepProps;
+    bulkActionsStepProps: EuiTourStepProps;
+  };
+  goToNextStep: () => void;
+  finishTour: () => void;
+}
+
+const STORAGE_KEY = 'securitySolution.rulesFeatureTour';
+const TOUR_POPOVER_WIDTH = 360;
+
+const featuresTourSteps: EuiStatelessTourStep[] = [
+  {
+    step: 1,
+    title: i18n.FEATURE_TOUR_IN_MEMORY_TABLE_STEP_TITLE,
+    content: <></>,
+    stepsTotal: 2,
+    children: <></>,
+    onFinish: noop,
+  },
+  {
+    step: 2,
+    title: i18n.FEATURE_TOUR_BULK_ACTIONS_STEP_TITLE,
+    content: <p>{i18n.FEATURE_TOUR_BULK_ACTIONS_STEP}</p>,
+    stepsTotal: 2,
+    children: <></>,
+    onFinish: noop,
+    anchorPosition: 'rightUp',
+  },
+];
+
+const tourConfig = {
+  currentTourStep: 1,
+  isTourActive: true,
+  tourPopoverWidth: TOUR_POPOVER_WIDTH,
+  tourSubtitle: i18n.FEATURE_TOUR_TITLE,
+};
+
+const RulesFeatureTourContext = createContext<RulesFeatureTourContextType | null>(null);
+
+/**
+ * Context for new rules features, displayed in demo tour(euiTour)
+ * It has a common state in useEuiTour, which allows transition from one step to the next, for components within it[context]
+ * It also stores tour's state in localStorage
+ */
+export const RulesFeatureTourContextProvider: FC = ({ children }) => {
+  const { storage } = useKibana().services;
+  const initialStore = useMemo(() => storage.get(STORAGE_KEY) ?? tourConfig, [storage]);
+
+  const [stepProps, actions, reducerState] = useEuiTour(featuresTourSteps, initialStore);
+
+  const finishTour = useCallback(() => actions.finishTour(), [actions]);
+  const goToNextStep = useCallback(() => actions.incrementStep(), [actions]);
+
+  const inMemoryTableStepProps = useMemo(
+    () => ({
+      ...stepProps[0],
+      maxWidth: TOUR_POPOVER_WIDTH,
+      content: (
+        <>
+          <p>{i18n.FEATURE_TOUR_IN_MEMORY_TABLE_STEP}</p>
+          <EuiSpacer />
+          <EuiButton color="primary" onClick={goToNextStep}>
+            {i18n.FEATURE_TOUR_IN_MEMORY_TABLE_STEP_NEXT}
+          </EuiButton>
+        </>
+      ),
+    }),
+    [stepProps, goToNextStep]
+  );
+  const bulkActionsStepProps = useMemo(
+    () => ({ ...stepProps[1], maxWidth: TOUR_POPOVER_WIDTH }),
+    [stepProps]
+  );
+
+  useEffect(() => {
+    storage.set(STORAGE_KEY, reducerState);
+  }, [reducerState, storage]);
```

Comment:
>  I would suggest introducing an advanced setting to disable the tour for the entire space.
Isn't setting work for all users in space? If one user disable it, others won't be able to see this tour at all.
Also, do all users have a right to rewrite adv. settings?

> Safari, for example, cleans local storage values if they are not accessed in a week.
I think we should be fine. That would a case only, if user don't open app for one week.

Do we have any other ways to keep settings for individual rules other than localStorage?
I remember, we discussed it couple of times - and localStorage seemed as valid and common solution for that case.
We also use localStorage for other features that requires per user settings, like experimental sorting toggle(the most recent one)

Reply:
IMHO we should be ok with using local storage in this case. If a user uses Safari and doesn't open Kibana for one week or more, it's not a big deal if they see the tour once again.

I think states like the tour's state, rules table experimental switch, etc are examples of a non-critical state that we can afford to lose. In other cases where we'd have critical state like user-generated data, we'd need to store it in Elasticsearch (e.g. saved objects). Here we might have a missing piece: individual user settings that would be stored in Elasticsearch per-user. I don't think such a thing exists in Kibana, but we don't have any requirements for it right now.

> We also use localStorage for other features that requires per user settings, like experimental sorting toggle(the most recent one)

Exactly, + persistent callouts store its state in local storage.

Reply:
Confirmed with the product. @jethr0null:

> let’s go ahead with it for this release. We should probably remove the callouts for 8.1's new features in 8,2 anyways so that gives us a release to talk to folks and figure out what the long term approach should be.

ES folks started working on a new "user profiles" feature that would allow persisting arbitrary data attached to a user (i.e. user preferences etc). When it is done and available as a Core Kibana service, we could stop using local storage and start persisting such data in Elasticsearch.